### PR TITLE
fix: add remark plugin to render raw HTML as literal text

### LINKED ISSUE #16417: Misc. bug: webui is dropping or omitting model's HTML output
### Name and Version

$ ./build/bin/llama-server --version
version: 6684 (606a73f5)
built with cc (Debian 14.2.0-19) 14.2.0 for x86_64-linux-gnu

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```shell
./build/bin/llama-server -m /mnt/models/DeepSeek-R1-0528-UD-IQ1_S-00001-of-00004.gguf --ctx-size 20480 --port 8080 --jinja --no-warmup --no-repack
./build/bin/llama-server -m /mnt/models/granite-4.0-h-small-Q6_K.gguf --ctx-size 16384 --port 8080 --jinja
./build/bin/llama-server -m /mnt/models/Qwen3-30B-A3B-Thinking-2507-Q4_K_M.gguf --ctx-size 16384 --port 8080 --jinja
```

### Problem description & steps to reproduce

I use the web UI to attach a ~30KB `.txt` file with my notes and instructions to create an HTML page with a table showing the results of my notes with some specific columns requested, as well as a summary and analysis.
For thinking models, the web UI correctly shows the reasoning tokens, but frequently drops parts or sometimes even entire final HTML output of the model. I can see that tokens are being produced, because the counters at the bottom of the page are going up, but nothing is showing up.
For non-thinking models some parts of the output might show up (like a single `<p>...</p>`) or nothing at all.
I also tested sending my query directly to the `/v1/chat/completions` API endpoint, and that endpoint produces a complete and correct response for the same model and server instance, no tokens missing.
I tested multiple different models - DeepSeek, Qwen, Granite - all of them had some issues, though some of them managed to produce some parts of the output in the web UI sometimes.
Based on the fact that this happens in multiple models on the web UI but the API does not have that problem, I assume there's a problem with handling HTML output in the UI.

### First Bad Commit

_No response_

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by ServeurpersoCom**
fix: add remark plugin to render raw HTML as literal text

Implemented a missing MDAST stage to neutralize raw HTML like major LLM WebUIs
do ensuring consistent and safe Markdown rendering

Introduced 'remarkLiteralHtml', a plugin that converts raw HTML nodes in the
Markdown AST into plain-text equivalents while preserving indentation and
line breaks. This ensures consistent rendering and prevents unintended HTML
execution, without altering valid Markdown structure

Kept 'remarkRehype' in the pipeline since it performs the required conversion
from MDAST to HAST for KaTeX, syntax highlighting, and HTML serialization

Refined the link-enhancement logic to skip unnecessary DOM rewrites,
fixing a subtle bug where extra paragraphs were injected after the first
line due to full innerHTML reconstruction, and ensuring links open in new
tabs only when required

Final pipeline: remarkGfm -> remarkMath -> remarkBreaks -> remarkLiteralHtml
-> remarkRehype -> rehypeKatex -> rehypeHighlight -> rehypeStringify

Close #16417

### CODE REVIEW

**Comment by ServeurpersoCom** [comment]
*Test sheet*

reasoning_content:
all must render as plain text

Final content:
markdown -> must render normally
markdown link -> must be clickable and open in a new tab
html without code block -> must render as plain text
html link tag -> must render as plain text and URL not clickable
html in code block -> must render as plain text + syntax-highlighted
latex in code block  -> must render as plain text
latex outside markdown -> must render as plain text
latex inside markdown (nominal LLM case) -> must render normally

This patch aligns the WebUI Markdown pipeline with industry-standard LLM renderers (OpenAI ChatGPT, Hugging Face Spaces, Anthropic...) by ensuring raw HTML safety without sacrificing formatting fidelity

This patch doesn't just "sanitize HTML" : it neutralizes raw XML-like output (e.g. \<think>, \<tool>, \<meta>, \<response>, \<step>, \<node>, \<data>), ensuring these symbolic or structural tags, whether produced by LLMs or part of generic XML fragments, are displayed as plain text rather than parsed as DOM, preserving structure while keeping the UI safe and consistent.

**Comment by ServeurpersoCom** [comment]
Sans titre

**Comment by zzokkolma** [comment]
I tested this PR and it seems to solve the missing HTML output part, however I did run into a formatting issue.
format-1

It is probably caused by the fact that the model output an empty indented line before that tag
format-2

**Comment by ServeurpersoCom** [comment]
Interesting case ! I just ran a quick test inside ChatGPT's own WebUI, and it fails in exactly the same way 😅
Even their renderer neutralizes raw HTML and collapses line breaks when HTML is escaped inside Markdown.
So what you're seeing is consistent with how most production-safe Markdown pipelines behave when remark-rehype or rehype-sanitize flatten the tree.

Here's a screenshot from that test:
GPT

I'll dig a bit deeper, but it really confirms that our remarkLiteralHtml stage is the right approach keeping structure visible while neutralizing unsafe tags.

**Comment by ServeurpersoCom** [comment]
To reproduce the issue now, you need to explicitly ask the model to output XML-like tags in the stream, which is already a bit of a hack, since LLMs naturally know they’re emitting Markdown.
So this goes slightly beyond normal usage, and as long as it’s only a rendering glitch, it’s probably better to keep the code simple rather than over-engineering for rare edge cases that could degrade the quality of conventional Markdown rendering.

**Comment by ServeurpersoCom** [comment]
Test prompt :

Write HTML with real blank lines and indentation inside a code block and then output the same HTML outside a code block, so we can compare the rendering.

https://chatgpt.com/share/68ea480b-5c3c-8012-9201-62cfb687dc67

And also on llama.cpp with this PR :

[conversation_6b69f066-c32c-4b0b-9f0d-92dad9c31764_tu_peux_crire_exacte.json](https://github.com/user-attachments/files/22863852/conversation_6b69f066-c32c-4b0b-9f0d-92dad9c31764_tu_peux_crire_exacte.json)

At this point, we’re actually doing slightly better than some major LLM WebUIs so that’s a good sign 😄
Let’s stop here before over-tuning it; the current behavior is safe, consistent, and covers all realistic use cases.

File: tools/server/webui/package-lock.json
Comment (allozaur) [inline tools/server/webui/package-lock.json:1]: Hmm, aren't we missing `package.json` changes in this PR?

We're importing these in literal HTML plugin:

```ts
import type { Plugin } from 'unified';
import { visit } from 'unist-util-visit';
import type { Break, Content, Paragraph, PhrasingContent, Root, Text } from 'mdast';
```

But I don't seee these packages installed in the diff.

File: tools/server/webui/src/lib/markdown/literalHtml.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	'emphasis',
+	'strong',
+	'delete',
+	'link',
+	'linkReference',
+	'tableCell'
+]);
+
+const NBSP = '\u00a0';
+const TAB_AS_SPACES = NBSP.repeat(4);
```
Comment (allozaur) [inline tools/server/webui/src/lib/markdown/literalHtml.ts]: I think it'd be great to move this to `$lib/constants` :)

File: tools/server/webui/src/lib/markdown/literal-html.ts
Comment (allozaur) [inline tools/server/webui/src/lib/markdown/literal-html.ts:1]: Let's rename this as `literal-html.ts`. We want to keep kebab-case for non-component file names.

**Review by allozaur — CHANGES_REQUESTED** [review]
Great stuff overall! Just left a few architectural remarks that need to be addressed.

File: tools/server/webui/package-lock.json
Comment (ServeurpersoCom) [reply tools/server/webui/package-lock.json:1]: Oops, you’re right : it still worked even after automatic "rm -rf node_modules" on my test server CI, because some other libraries were already pulling in those same dependencies 😅

Comment (ServeurpersoCom) [reply tools/server/webui/package-lock.json:1]: On master I get this warning :

Sans titre

Comment (ServeurpersoCom) [reply tools/server/webui/package-lock.json:1]: Done

File: tools/server/webui/src/lib/markdown/literalHtml.ts (hunk 1 above)
Comment (ServeurpersoCom) [reply tools/server/webui/src/lib/markdown/literalHtml.ts]: Done

File: tools/server/webui/src/lib/markdown/literal-html.ts
Comment (ServeurpersoCom) [reply tools/server/webui/src/lib/markdown/literal-html.ts:1]: Done !

File: tools/server/public/index.html.gz
Comment (allozaur) [inline tools/server/public/index.html.gz:1]: @ServeurpersoCom let's update the static build and ship this, as it'd be good to have this merged first before moving forward with #16496 and #16508

**Review by allozaur — CHANGES_REQUESTED** [review]

File: tools/server/public/index.html.gz
Comment (ServeurpersoCom) [reply tools/server/public/index.html.gz:1]: Done : rebase + format (already done) + static build updated + pushed

Comment (allozaur) [reply tools/server/public/index.html.gz:1]: Sweet!

**Review by allozaur — APPROVED** [review]
